bde/ml/models.py

```diff
@@ -827,17 +897,137 @@ def _prior_fitting(
             static_broadcasted_argnums=[2],
         )(
             model_states,
-            jnp.arange(self.n_init_runs),
+            jnp.arange(self.epochs),
             self.loss,
             jnp.array(metrics),
             train,
             valid,
         )
         return params, history
 
-    def mcmc_loop(self):
-        r"""Perform MCMC sampling for BDE estimator."""
-        ...
+    def mcmc_sampling(
+        self,
+        model_states: train_state.TrainState,
+        rng_key: PRNGKeyArray,
+        train: datasets.BasicDataset,
+    ) -> List[Dict]:
+        r"""Perform MCMC-burn-in and sampling."""
+        sample_keys = jax.random.split(rng_key, self.n_chains + 1)
+        split_key, sample_keys = sample_keys[0], sample_keys[1:]
+
+        @jax.jit
+        def logdensity_for_batch(
+            params,
+            carry: float,
+            batch: Tuple[ArrayLike, ArrayLike],
+        ) -> Tuple[float, None]:
+            x, y = batch
+            y_pred = self.model_.apply(params, x)
+            res_loss = self.loss(y, y_pred)
+            res_loss = jnp.sum(res_loss)
+            res_prior = self.log_prior(params)
```

Comment:
should not be applied n_batches many times! We do a full batch evaluation meaning we push the whole dataset at once. So we indeed can du a running sum of the log likelihood but not over the log prior over the params which are independent of the batch/dataset!

Also the loss is only proportional to the log likelihood please use the exact likelihood here. also adjust the docstring in the loss definition saying it calculates the log likelihood. it does not - it just returns something proportional to it.

Reply:
also you apply log to the learned std - shouldn't it be the exp?

Reply:
Does applying the data in batches break the algorithm?
If it doesn't I would rather keep it since it can reduce the resource requirements.
Maybe we could keep the default batch_size value as -1 to indicate a full batch evaluation and give the user the ability to change it, since the batching is already implemented.

Reply:
and does the full batch requirement apply only to the mcmc-sampling, or to the DE-init as well?

Reply:
Batching definitely breaks the sampling algorithm yes. The user should not be able to change this as of right now with this sampler. DE warmstarts can and should be trained in a vanilla batched way.

Reply:
even if it is always is single batch this is not how code should be written. It is tempting to extend it to minibatch samplers and then it will be wrong

Reply:
Right now I hardcoded `self.batch_mcmc_ = X.shape[0]` in the `fit` method, and the user has no way to change it, so the minibatch algorithm is ran on only 1 batch (batch_size = n_data_points -> 1 iter).
If this solution works with MCMC sampling I would like to keep it, since:
 - If the sampling method will be changed in the future to one which allows for minibatching, we will have the code ready to go.
 - I don't think this should effect efficiency in a noticeable way
 
@EmanuelSommer what do you think?

Reply:
It's ok like this, but as soon as mini batches are used the code will be methodologically wrong due to the prior. This is unfortunate.